test/services/MessageService.spec.ts

```diff
@@ -0,0 +1,56 @@
+import Message from '@/interfaces/Message';
+import MessageModel from '@/models/message';
+import MessageService from '@/services/MessageService';
+import message from '@test/data/message';
+import { AssertionError } from 'assert';
+import chai from 'chai';
+import spies from 'chai-spies';
+import subset from 'chai-subset';
+import 'mocha';
+import mongoose from 'mongoose';
+
+// Add extensions to chai
+chai.use(spies);
+chai.use(subset);
+const expect = chai.expect;
+
+describe('MessageService.send', () => {
+
+  // Instantiate the service
+  const messageService = new MessageService();
+
+  before('should establish a database connection.', (done) => {
+    // connect to database
+    const db = mongoose.connection;
+    db.on('error', console.error.bind(console, 'connection error:'));
+    db.once('open', done);
+    mongoose.connect(`mongodb://${process.env.MONGO_HOST || 'localhost'}/notification-service-test`);
+  });
+
+  it('should call the escalation logic.', async () => {
+    const spyFunction = chai.spy();
+    (messageService as any).escalationLogic.escalate = spyFunction;
+
+    await messageService.send(message);
+
+    expect(spyFunction)
+      .to.have.been.called();
+  });
+
+  it('should write the message to the database.', async () => {
+    const messageId = await messageService.send(message);
+    const databaseMessageModel = await MessageModel.findById(messageId);
+    if (!databaseMessageModel) {
```

Comment:
I am not sure whether throwing assertions is ideal in a test. Maybe we could use something like this: `expect(databaseMessageModel, 'Could not find message in database.').to.exist;`

A custom error message can be given as the second argument to `expect `.

Reply:
Changed this to 
```
if (!databaseMessageModel) {
      expect(databaseMessageModel, 'Could not find message in database.').not.to.be.null;
      return;
}
```

Reply:
Why do we need the if condition and the return statement here? If `!databaseMessageModel` evaluates to `true` then also `expect(databaseMessageModel, 'Could not find message in database.').not.to.be.null;` will fail, right?

Reply:
Yes, it will but the code will also be executed further (without that conditional return) and may run into an exception while accessing the databaseMessageModel. That's why the conditional return is required at this point.